src/envoy/http/authn/http_filter_test.cc

```diff
@@ -0,0 +1,144 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/authn/http_filter.h"
+#include "authentication/v1alpha1/policy.pb.h"
+#include "common/http/header_map_impl.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "src/envoy/http/authn/authenticator_base.h"
+#include "test/mocks/http/mocks.h"
+
+using Envoy::Http::Istio::AuthN::AuthenticatorBase;
+using Envoy::Http::Istio::AuthN::FilterContext;
+using testing::_;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Http {
+namespace {
+
+// Create a fake authenticator for test. This authenticator do nothing except
+// making done callback with failure.
+AuthenticatorBase* createAlwaysFailAuthenticator(
+    FilterContext* filter_context,
+    const AuthenticatorBase::DoneCallback& done_callback) {
+  class _local : public AuthenticatorBase {
+   public:
+    _local(FilterContext* filter_context,
+           const AuthenticatorBase::DoneCallback& callback)
+        : AuthenticatorBase(filter_context, callback) {}
+    void run() override { done(false); }
+  };
+  return new _local(filter_context, done_callback);
+}
+
+// Create a fake authenticator for test. This authenticator do nothing except
+// making done callback with success.
+AuthenticatorBase* createAlwaysPassAuthenticator(
```

Comment:
This can be easily done by using MockAuthenticatorBase  and mock run() method, and have EXPECT_CALL on the run() method to return true/false

Reply:
Not exactly. The http_filter own and hide these objects, you don't have access to them to hook EXPECT_CALL without writing more complex test code.